sigma_fun/src/ext/dl_secp256k1_ed25519_eq.rs

```diff
@@ -0,0 +1,248 @@
+use crate::{
+    ed25519,
+    or::Either,
+    secp256k1,
+    secp256k1::fun::{
+        g,
+        marker::*,
+        rand_core::{CryptoRng, RngCore},
+        s, Point as PointP, Scalar as ScalarP, G as GP,
+    },
+    All, And, Eq, FiatShamir, Or, Sigma, Transcript,
+};
+use alloc::vec::Vec;
+use curve25519_dalek::{
+    constants::ED25519_BASEPOINT_TABLE, edwards::EdwardsPoint as PointQ, scalar::Scalar as ScalarQ,
+    traits::Identity,
+};
+use generic_array::typenum::{U252, U31};
+static GQ: &'static curve25519_dalek::edwards::EdwardsBasepointTable = &ED25519_BASEPOINT_TABLE;
+
+/// The underlying proof algorithm we'll be using to prove the relationship between the commitments and the keys.
+type CoreProof = And<
+    All<
+        U252, // For each of the 252 bits of the secret key
+        // We show that both commitments are a commitment to zero OR to 2^i for i = 0..252
+        Or<
+            And<secp256k1::DLBP<U31>, ed25519::DLBP<U31>>,
+            And<secp256k1::DLBP<U31>, ed25519::DLBP<U31>>,
+        >,
+    >,
+    // Finally we prove the result of the addition of the commitments is the same as the ones calimed
+    // i.e. if the commitmens add up to xH, we show that X = xG.
+    And<Eq<secp256k1::DLBP<U31>, secp256k1::DL<U31>>, Eq<ed25519::DLBP<U31>, ed25519::DL<U31>>>,
+>;
+const COMMITMENT_BITS: usize = 252;
+
+#[derive(Debug, Clone)]
+pub struct CrossCurveDLEQProof {
+    claim: (PointP, PointQ),
+    sum_blindings: (ScalarP<Public, Zero>, ScalarQ),
+    commitments: Vec<(PointP, PointQ)>,
+    proof: crate::CompactProof<CoreProof>,
+}
+
+pub struct CrossCurveDLEQ<T> {
+    HQ: PointQ,
+    HP: PointP,
+    core_proof_system: FiatShamir<CoreProof, T>,
+    powers_of_two: Vec<(PointP, PointQ)>,
+}
+
+impl<T: Transcript<CoreProof>> CrossCurveDLEQ<T> {
+    pub fn new(HP: PointP, HQ: PointQ) -> Self {
+        let powers_of_two = core::iter::successors(Some((HP.clone(), HQ.clone())), |(H2P, H2Q)| {
+            // compute 2^i * H for i = 0..252 by successively adding the result of the last addition
+            Some((
+                g!(H2P + H2P).mark::<(Normal, NonZero)>().unwrap(),
+                (H2Q + H2Q),
+            ))
+        })
+        .take(COMMITMENT_BITS)
+        .collect();
+
+        Self {
+            HP,
+            HQ,
+            core_proof_system: FiatShamir::new(CoreProof::default()),
+            powers_of_two,
+        }
+    }
+
+    /// Generates the two corresponding points for the same 252-bit ed25519
+    /// secret and generates a proof that they have the same discrete logarithm.
+    ///
+    /// # Panics
+    ///
+    /// - If the secret is larger than 2^253 -1
+    /// - If the secret is 0
+    pub fn prove(
+        &self,
+        secret: &ScalarQ,
+        rng: &mut (impl CryptoRng + RngCore),
+    ) -> CrossCurveDLEQProof {
+        // Must be a 252 bit ed25519 key
+        assert!(secret.as_bytes()[31] & 0x20 == 0);
```

Comment:
I'm a real newbie when it comes to bitwise operations, but I think that this only proves that the scalar is 255 bits, not 252.